### ISSUE TITLE
Feature suggestion: quick prototype tts files

### ISSUE THREAD

**Issue by quetzalsly** [quisquous/cactbot#489]
Today eden savage came out and I had a cool idea, what if we could quickly prototype fights without waiting for cactbot to update? and then possibly also helping with pull requests.

The timeline and trigger system file feels very bloated to me, it takes so long to make something like this for the average person who is a non programmer.

My proposal is a very small file which you have open in notepad++ and on each change just reload the overlay and it reloads everything nice and easy.

There would be a new folder for these files and each separate boss would have their own file. The file can be something like this:

```
//Would use javascript but I just wrote this as example pseudocode
//Non regex for average person to easily make triggers!
//Totally independant from any timeline system, its just raw tts

//Personal variables
const myOrb = "West";
//

version: 0.1,
author: "Name",
zone: "Eden's Gate: Resurrection (Savage)",
boss: "Eden Prime",
array
{
    cast1: {0. "gravity", "heal aoe"}, //ms delay before announcement, string match (case 
    incensitive), tts (I guess this can be any language?)
    cast2: {500, "flare", "go in"},
    cast3: {200, "something", "get ready for orbs"},
    cast4: {0, "something", "go to " + myOrb + " orb!"}
}
```

**Comment by DrLippe** [comment]
```js
{
  regex: /:Eden Prime starts using Gravity/,
  delay: 0,
  tts: "heal aoe"
}
```
This is your `cast1` as a trigger. You could simply add this to your custom triggers and done.
The problem with your approach is, that your string match is way too simple. It would trigger every time `gravity` shows up in log, which is -> when boss starts using it, when he used it, for everyone that got hit by it, someone talking about gravity in the chat etc.
Since you really need a proper regex, writing the trigger in javascript like above is the easiest approach.

What I would suggest is linking the source of your raidboss overlay to the online version. This way you wont need to update your cactbot to access new trigger files.

**Comment by quetzalsly** [comment]
Hmm yeah I thought maybe behind the scenes matching the string to boss name, but yeah your way makes perfect sense. Can you please explain what you mean linking the overlay to online version? Basically all I wanted to achieve with my idea was a very simple way to make these files. I am very apreciative that the author updated eden savage on the first day, but I am thinking what if the author stops raiding and no one would be updating it anymore. If there was a simple format people could also very easily share it on discord and import it instantly without doing a bunch of copy pasting with cactbot files.

**Comment by DrLippe** [comment]
There is a documentation on how to create trigger files yourself. [cactbot/blob/master/AdvancedCactbot.md#adding-custom-triggers](url)
Also you can contribute to this repository.

You could link the overlay to [https://quisquous.github.io/cactbot/ui/raidboss/raidboss.html](url).
But since this is under constant development, the overlay may not work all the time.

Another way of simplifying the updating is by cloning the repository to your local files and sync.

**Comment by JLGarber** [comment]
Unfortunately there's not an easy way to match battle effects 1-to-1 with human-readable text in the fashion presented in the feature request.  Every enemy ability in the game has a unique ID associated with it, a hexadecimal number in the range 0000-FFFF. (This is equivalent to 0-65,535 in decimal, although the vast majority of these numbers are currently unused.) Cactbot matches enemy abilities against these IDs as well as the cast names for the precise reason DrLippe notes: If the addon just matched against the name, there would be tons of false positives. Also, there are situations where the boss casts an ability with a given name, and then uses an ability with the same exact human-readable name but a different ID. With just text it's impossible to separate these.

In addition to that, there are a number of battle effects that aren't actually cast from the boss on a target. Taking Eden 1 normal or savage as an example, the first Vice/Virtue cast isn't targeted at the damage dealers, but at Eden itself. The effect then just appears on players. You wouldn't be able to model this without digging into the ACT log and discovering that.

Long story short, the benefit from Quisquous or someone else adding this kind of utility would be limited to people who A) are doing serious progression on the first day and B) are part of a raid team that doesn't have at least one person with the minimal programming knowledge to create triggers. That's an extremely small group of people to make use of something that would be a reasonably significant amount of work.

That being said, it's not terribly difficult to look for these abilities on your own and write triggers based on that. Have a look at https://github.com/quisquous/cactbot/blob/master/docs/LogGuide.md and see what you think.

**Comment by quetzalsly** [comment]
Thanks for explaining this JLGarber. I did not know the system was so complex. To quisquous: thanks so much for updating this so fast! I am running e2s and it's like 95% perfect tts, not sure why it's called 'protean' though :) Closing issue cos it's no longer an issue.

**Comment by JLGarber** [comment]
@quetzalsly Happy to help. As for "Protean", that's the "traditional" name. That mechanic first appeared in Heavensward, in Alexander 3. ([An example from Mizzteq's old guide video.](https://youtu.be/21yhP09YLUo?t=50)) It's appeared in slightly different forms elsewhere, but that's the "universal" name that everyone gives it. (You might be familiar with it as Queen's Waltz from Omega 3, or Flame Thrower on Omega 11.)

**Comment by quisquous** [comment]
Sorry for calling it protean when maybe that doesn't make sense, but it was hard to come up with something succinct to say.  At least I don't call every floor aoe as its traditional "plumes" name? :wink:

Thanks for answering this for me.  I think most folks would be better off learning a little bit about regular expressions to be able to write the full version of the triggers in the first place.
